<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>bbcflib.mapseq</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="bbcflib-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="http://bbcf.epfl.ch/bbcflib">bbcflib</a></th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="bbcflib-module.html">Package&nbsp;bbcflib</a> ::
        Module&nbsp;mapseq
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="bbcflib.mapseq-module.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== MODULE DESCRIPTION ==================== -->
<h1 class="epydoc">Module mapseq</h1><p class="nomargin-top"></p>
<pre class="literalblock">

======================
Module: bbcflib.mapseq
======================

This module provides functions useful to map raw reads using bowtie.
The most common workflow will first use ``map_reads`` which takes the following arguments:

  * ``'ex'``: an execution environment to run jobs in,

  * ``'fastq_file'``: the raw reads as a fastq file,

  * ``'chromosomes'``: a dictionary with keys 'chromosome_id' as used in the bowtie indexes and values a dictionary with usual chromosome names and their lengths,

  * ``'bowtie_index'``: the file prefix to the bowtie index,

  * ``'maxhits'``: the maximum number of hits per read to keep in the output (default *5*),

  * ``'antibody_enrichment'``: an approximate enrichement ratio of protein bound sequences over controls (default *50*),

  * ``'name'``: sample name to be used in descriptions and file names,

  * ``'remove_pcr_duplicates'``: whether to remove probable PCR amplification artifacts based on a Poisson confidence threshold (default *True*),

  * ``'bwt_args'``: a list of specific arguments for bowtie (in addition to [&quot;-Sam&quot;,str(max(20,maxhits)),&quot;--best&quot;,&quot;--strata&quot;]).

The function ``map_groups`` will take a collection of sample as described in a *job* object from the ``frontend`` module and run fetch fastq files for each of them through using a ``daflims`` object, use an 'Assembly' object from ``genrep`` to get the bowtie indexes and run ``map_reads`` for each sample.

The second step in the workflow consists in generating genome-wide density maps with ``densities_groups`` which needs a 'bein' execution environment, the same *job* and *assembly* as above and a dictionary 'file_dict' as returned by ``map_groups``. The function then runs ``parallel_density_sql`` on each 'bam' file to obtain a normalized read density profile as a 'sqlite' file.

Below is the script used by the frontend::

    from bbcflib import daflims, genrep, frontend, gdv, common
    from bbcflib.mapseq import *
    M = MiniLIMS( limspath )
    working_dir = '/path/to/scratch/on/cluster'
    hts_key = 'test_key'
    gl = { 'hts_mapseq': {'url': 'http://htsstation.vital-it.ch/mapseq/'},
           'genrep_url': 'http://bbcftools.vital-it.ch/genrep/',
           'bwt_root': '/db/genrep/nr_assemblies',
           'script_path': '/srv/chipseq/lib',
           'lims': {'user': 'alice',
                     'passwd': {'lgtf': 'bob123',
                                'gva': 'joe456'}},
            'gdv': {'url': 'http://svitsrv25.epfl.ch/gdv',
                    'email': 'alice.ecila@somewhere.edu',
                    'key': 'xxxxxxxxxxxxxxxxxxxxxxxx'} }
    assembly_id = 'mm9'
    htss = frontend.Frontend( url=gl['hts_mapseq']['url'] )
    job = htss.job( hts_key )
    g_rep = genrep.GenRep( gl['genrep_url'], gl['bwt_root'] )
    assembly = g_rep.assembly( assembly_id )
    daflims1 = dict((loc,daflims.DAFLIMS( username=gl['lims']['user'],
                                          password=gl['lims']['passwd'][loc] ))
                    for loc in gl['lims']['passwd'].keys())
    job.options['ucsc_bigwig'] = True
    with execution( M, description=hts_key, remote_working_directory=working_dir ) as ex:
        job = get_fastq_files( job, ex.working_directory, daflims1 )
        mapped_files = map_groups( ex, job, ex.working_directory, assembly )
        pdf = add_pdf_stats( ex, mapped_files,
                             dict((k,v['name']) for k,v in job.groups.iteritems()),
                             gl['script_path'] )
        density_files = densities_groups( ex, job, mapped_files, assembly.chromosomes )
        gdv_project = gdv.create_gdv_project( gl['gdv']['key'], gl['gdv']['email'],
                                              job.description, hts_key,
                                              assembly.nr_assembly_id,
                                              gdv_url=gl['gdv']['url'],
                                              public=True )
        add_pickle( ex, gdv_project, description='py:gdv_json' )
    print ex.id
    allfiles = common.get_files( ex.id, M )
    print allfiles

</pre>

<!-- ==================== FUNCTIONS ==================== -->
<a name="section-Functions"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Functions</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Functions"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_compact_chromosome_name"></a><span class="summary-sig-name">_compact_chromosome_name</span>(<span class="summary-sig-arg">key</span>)</span></td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.mapseq-module.html#add_pdf_stats" class="summary-sig-name">add_pdf_stats</a>(<span class="summary-sig-arg">ex</span>,
        <span class="summary-sig-arg">processed</span>,
        <span class="summary-sig-arg">group_names</span>,
        <span class="summary-sig-arg">script_path</span>,
        <span class="summary-sig-arg">description</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">pdf:mapping_report.pdf</code><code class="variable-quote">'</code></span>)</span><br />
      Runs the 'plot_stats' function and adds its pdf output to the 
      execution's repository.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.mapseq-module.html#densities_groups" class="summary-sig-name">densities_groups</a>(<span class="summary-sig-arg">ex</span>,
        <span class="summary-sig-arg">job_or_dict</span>,
        <span class="summary-sig-arg">file_dict</span>,
        <span class="summary-sig-arg">chromosomes</span>,
        <span class="summary-sig-arg">via</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">lsf</code><code class="variable-quote">'</code></span>)</span><br />
      Arguments are:</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.mapseq-module.html#get_bam_wig_files" class="summary-sig-name">get_bam_wig_files</a>(<span class="summary-sig-arg">ex</span>,
        <span class="summary-sig-arg">job</span>,
        <span class="summary-sig-arg">minilims</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">hts_url</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">suffix</span>=<span class="summary-sig-default"><code class="variable-group">[</code><code class="variable-quote">'</code><code class="variable-string">fwd</code><code class="variable-quote">'</code><code class="variable-op">, </code><code class="variable-quote">'</code><code class="variable-string">rev</code><code class="variable-quote">'</code><code class="variable-group">]</code></span>,
        <span class="summary-sig-arg">script_path</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">./</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">via</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">lsf</code><code class="variable-quote">'</code></span>)</span><br />
      Will replace file references by actual file paths in the 'job' 
      object.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.mapseq-module.html#get_fastq_files" class="summary-sig-name">get_fastq_files</a>(<span class="summary-sig-arg">job</span>,
        <span class="summary-sig-arg">fastq_root</span>,
        <span class="summary-sig-arg">dafl</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">set_seed_length</span>=<span class="summary-sig-default">True</span>)</span><br />
      Will replace file references by actual file paths in the 'job' 
      object.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.mapseq-module.html#import_mapseq_results" class="summary-sig-name">import_mapseq_results</a>(<span class="summary-sig-arg">key_or_id</span>,
        <span class="summary-sig-arg">minilims</span>,
        <span class="summary-sig-arg">ex_root</span>,
        <span class="summary-sig-arg">url_or_dict</span>)</span><br />
      Imports all files created by a previous 'mapseq' workflow into the 
      current execution environement.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.mapseq-module.html#map_groups" class="summary-sig-name">map_groups</a>(<span class="summary-sig-arg">ex</span>,
        <span class="summary-sig-arg">job_or_dict</span>,
        <span class="summary-sig-arg">fastq_root</span>,
        <span class="summary-sig-arg">assembly_or_dict</span>,
        <span class="summary-sig-arg">map_args</span>=<span class="summary-sig-default">None</span>)</span><br />
      Fetches fastq files and bowtie indexes, and runs the 'map_reads' 
      function for a collection of samples described in a 'Frontend' 'job'.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.mapseq-module.html#map_reads" class="summary-sig-name">map_reads</a>(<span class="summary-sig-arg">ex</span>,
        <span class="summary-sig-arg">fastq_file</span>,
        <span class="summary-sig-arg">chromosomes</span>,
        <span class="summary-sig-arg">bowtie_index</span>,
        <span class="summary-sig-arg">maxhits</span>=<span class="summary-sig-default">5</span>,
        <span class="summary-sig-arg">antibody_enrichment</span>=<span class="summary-sig-default">50</span>,
        <span class="summary-sig-arg">name</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string"></code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">remove_pcr_duplicates</span>=<span class="summary-sig-default">True</span>,
        <span class="summary-sig-arg">bwt_args</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">via</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">lsf</code><code class="variable-quote">'</code></span>)</span><br />
      Runs ``bowtie`` in parallel over lsf for the `fastq_file` input.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.mapseq-module.html#parallel_density_sql" class="summary-sig-name">parallel_density_sql</a>(<span class="summary-sig-arg">ex</span>,
        <span class="summary-sig-arg">bamfile</span>,
        <span class="summary-sig-arg">chromosomes</span>,
        <span class="summary-sig-arg">nreads</span>=<span class="summary-sig-default">1</span>,
        <span class="summary-sig-arg">merge</span>=<span class="summary-sig-default">-1</span>,
        <span class="summary-sig-arg">read_extension</span>=<span class="summary-sig-default">-1</span>,
        <span class="summary-sig-arg">convert</span>=<span class="summary-sig-default">True</span>,
        <span class="summary-sig-arg">b2w_args</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">via</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">lsf</code><code class="variable-quote">'</code></span>)</span><br />
      Runs 'bam_to_density' for every chromosome in the 'chromosomes' list.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.mapseq-module.html#parallel_density_wig" class="summary-sig-name">parallel_density_wig</a>(<span class="summary-sig-arg">ex</span>,
        <span class="summary-sig-arg">bamfile</span>,
        <span class="summary-sig-arg">chromosomes</span>,
        <span class="summary-sig-arg">nreads</span>=<span class="summary-sig-default">1</span>,
        <span class="summary-sig-arg">merge</span>=<span class="summary-sig-default">-1</span>,
        <span class="summary-sig-arg">read_extension</span>=<span class="summary-sig-default">-1</span>,
        <span class="summary-sig-arg">convert</span>=<span class="summary-sig-default">True</span>,
        <span class="summary-sig-arg">description</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string"></code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">alias</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">b2w_args</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">via</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">lsf</code><code class="variable-quote">'</code></span>)</span><br />
      Runs 'bam_to_density' in parallel for every chromosome in the 
      'chromosomes' list with 'sql' set to False.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.mapseq-module.html#poisson_threshold" class="summary-sig-name">poisson_threshold</a>(<span class="summary-sig-arg">mu</span>,
        <span class="summary-sig-arg">cutoff</span>=<span class="summary-sig-default">0.95</span>,
        <span class="summary-sig-arg">max_terms</span>=<span class="summary-sig-default">100</span>)</span><br />
      Calculate confidence threshold for Poisson distributions.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.mapseq-module.html#pprint_bamstats" class="summary-sig-name">pprint_bamstats</a>(<span class="summary-sig-arg">sample_stats</span>)</span><br />
      Pretty stdout-print for sample_stats.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.mapseq-module.html#remove_duplicate_reads" class="summary-sig-name">remove_duplicate_reads</a>(<span class="summary-sig-arg">bamfile</span>,
        <span class="summary-sig-arg">chromosomes</span>,
        <span class="summary-sig-arg">maxhits</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">pilesize</span>=<span class="summary-sig-default">1</span>,
        <span class="summary-sig-arg">convert</span>=<span class="summary-sig-default">False</span>)</span><br />
      Filters a bam file for multi-hits above 'maxhits' and for duplicate 
      reads beyond 'pilesize'.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
</table>
<!-- ==================== VARIABLES ==================== -->
<a name="section-Variables"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Variables</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Variables"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="__package__"></a><span class="summary-name">__package__</span> = <code title="'bbcflib'"><code class="variable-quote">'</code><code class="variable-string">bbcflib</code><code class="variable-quote">'</code></code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="bam_to_density"></a><span class="summary-name">bam_to_density</span> = <code title="&lt;bein.program object at 0x10d425810&gt;">&lt;bein.program object at 0x10d425810&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="bam_to_sam"></a><span class="summary-name">bam_to_sam</span> = <code title="&lt;bein.program object at 0x10c04bc10&gt;">&lt;bein.program object at 0x10c04bc10&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="bamstats"></a><span class="summary-name">bamstats</span> = <code title="&lt;bein.program object at 0x10d425310&gt;">&lt;bein.program object at 0x10d425310&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="bowtie"></a><span class="summary-name">bowtie</span> = <code title="&lt;bein.program object at 0x10c04bad0&gt;">&lt;bein.program object at 0x10c04bad0&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="bowtie_build"></a><span class="summary-name">bowtie_build</span> = <code title="&lt;bein.program object at 0x10c04bb50&gt;">&lt;bein.program object at 0x10c04bb50&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="count_lines"></a><span class="summary-name">count_lines</span> = <code title="&lt;bein.program object at 0x10c04ba10&gt;">&lt;bein.program object at 0x10c04ba10&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="exp"></a><span class="summary-name">exp</span> = <code title="&lt;ufunc 'exp'&gt;">&lt;ufunc 'exp'&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="external_add_nh_flag"></a><span class="summary-name">external_add_nh_flag</span> = <code title="&lt;bein.program object at 0x10c04bb90&gt;">&lt;bein.program object at 0x10c04bb90&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="index_bam"></a><span class="summary-name">index_bam</span> = <code title="&lt;bein.program object at 0x10c04bd50&gt;">&lt;bein.program object at 0x10c04bd50&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="md5sum"></a><span class="summary-name">md5sum</span> = <code title="&lt;bein.program object at 0x10c04b990&gt;">&lt;bein.program object at 0x10c04b990&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="merge_bam"></a><span class="summary-name">merge_bam</span> = <code title="&lt;bein.program object at 0x10c04bd90&gt;">&lt;bein.program object at 0x10c04bd90&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="plot_stats"></a><span class="summary-name">plot_stats</span> = <code title="&lt;bein.program object at 0x10d425590&gt;">&lt;bein.program object at 0x10d425590&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="remove_lines_matching"></a><span class="summary-name">remove_lines_matching</span> = <code title="&lt;bein.program object at 0x10c04b950&gt;">&lt;bein.program object at 0x10c04b950&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="replace_bam_header"></a><span class="summary-name">replace_bam_header</span> = <code title="&lt;bein.program object at 0x10c04bc50&gt;">&lt;bein.program object at 0x10c04bc50&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="sam_to_bam"></a><span class="summary-name">sam_to_bam</span> = <code title="&lt;bein.program object at 0x10c04bbd0&gt;">&lt;bein.program object at 0x10c04bbd0&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="sleep"></a><span class="summary-name">sleep</span> = <code title="&lt;bein.program object at 0x10c04b9d0&gt;">&lt;bein.program object at 0x10c04b9d0&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="sort_bam"></a><span class="summary-name">sort_bam</span> = <code title="&lt;bein.program object at 0x10c04bc90&gt;">&lt;bein.program object at 0x10c04bc90&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="sort_bam_by_read"></a><span class="summary-name">sort_bam_by_read</span> = <code title="&lt;bein.program object at 0x10c04bcd0&gt;">&lt;bein.program object at 0x10c04bcd0&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="split_file"></a><span class="summary-name">split_file</span> = <code title="&lt;bein.program object at 0x10c04ba50&gt;">&lt;bein.program object at 0x10c04ba50&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="touch"></a><span class="summary-name">touch</span> = <code title="&lt;bein.program object at 0x10c04b910&gt;">&lt;bein.program object at 0x10c04b910&gt;</code>
    </td>
  </tr>
</table>
<!-- ==================== FUNCTION DETAILS ==================== -->
<a name="section-FunctionDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Function Details</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-FunctionDetails"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
</table>
<a name="add_pdf_stats"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">add_pdf_stats</span>(<span class="sig-arg">ex</span>,
        <span class="sig-arg">processed</span>,
        <span class="sig-arg">group_names</span>,
        <span class="sig-arg">script_path</span>,
        <span class="sig-arg">description</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">pdf:mapping_report.pdf</code><code class="variable-quote">'</code></span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Runs the 'plot_stats' function and adds its pdf output to the 
  execution's repository.</p>
  <p>Arguments are the output of 'map_groups' ('processed'), a dictionary 
  of group_id to names used in the display, the path to the script used by 
  'plot_stats', and the 'description' to use in the repository.</p>
  <p>Returns the name of the pdf file.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="densities_groups"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">densities_groups</span>(<span class="sig-arg">ex</span>,
        <span class="sig-arg">job_or_dict</span>,
        <span class="sig-arg">file_dict</span>,
        <span class="sig-arg">chromosomes</span>,
        <span class="sig-arg">via</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">lsf</code><code class="variable-quote">'</code></span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Arguments are:</p>
  <p>* ``'ex'``: a 'bein' execution environment to run jobs in,</p>
  <p>* ``'job_or_dict'``: a 'Frontend' 'job' object, or a dictionary with 
  keys 'groups',</p>
  <p>* ``'file_dict'``: a dictionary of files,</p>
  <p>* ``'chromosomes'``: a dictionary with keys 'chromosome_id' as used in
  the bowtie indexes and values a dictionary with usual chromosome names 
  and their lengths,</p>
  <p>Returns a dictionary with keys *group_id* from the job object and 
  values the files fo each group ('bam' and 'wig').</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_bam_wig_files"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_bam_wig_files</span>(<span class="sig-arg">ex</span>,
        <span class="sig-arg">job</span>,
        <span class="sig-arg">minilims</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">hts_url</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">suffix</span>=<span class="sig-default"><code class="variable-group">[</code><code class="variable-quote">'</code><code class="variable-string">fwd</code><code class="variable-quote">'</code><code class="variable-op">, </code><code class="variable-quote">'</code><code class="variable-string">rev</code><code class="variable-quote">'</code><code class="variable-group">]</code></span>,
        <span class="sig-arg">script_path</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">./</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">via</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">lsf</code><code class="variable-quote">'</code></span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Will replace file references by actual file paths in the 'job' object.
  These references are either 'mapseq' keys or urls.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_fastq_files"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_fastq_files</span>(<span class="sig-arg">job</span>,
        <span class="sig-arg">fastq_root</span>,
        <span class="sig-arg">dafl</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">set_seed_length</span>=<span class="sig-default">True</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Will replace file references by actual file paths in the 'job' object.
  These references are either 'dafl' run descriptions or urls. Argument 
  'dafl' is a dictionary of 'Daflims' objects (keys are the facility 
  names). If 'set_seed_length' is true, a dictionary 
  job.groups[gid]['seed_lengths'] of seed lengths is constructed with 
  values corresponding to 70% of the read length.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="import_mapseq_results"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">import_mapseq_results</span>(<span class="sig-arg">key_or_id</span>,
        <span class="sig-arg">minilims</span>,
        <span class="sig-arg">ex_root</span>,
        <span class="sig-arg">url_or_dict</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Imports all files created by a previous 'mapseq' workflow into the 
  current execution environement.</p>
  <p>* ``'key_or_id'`` is the previous execution's description in the 
  MiniLIMS or its id,</p>
  <p>* ``'minilims'`` is the MiniLIMS where that execution was saved,</p>
  <p>* ``'ex_root'`` is the current execution's directory (where files will
  be copied to),</p>
  <p>* ``'url_or_dict'`` is either the 'Frontend' url to fetch the mapseq 
  job's description, or a job description dictionary (see 
  ``map_groups``)</p>
  <p>Returns a tuple with a dictionary similar to the output of 
  'map_groups' and a 'job' object describing the mapseq runs.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="map_groups"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">map_groups</span>(<span class="sig-arg">ex</span>,
        <span class="sig-arg">job_or_dict</span>,
        <span class="sig-arg">fastq_root</span>,
        <span class="sig-arg">assembly_or_dict</span>,
        <span class="sig-arg">map_args</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Fetches fastq files and bowtie indexes, and runs the 'map_reads' 
  function for a collection of samples described in a 'Frontend' 'job'.</p>
  <p>Arguments are:</p>
  <p>* ``'ex'``: a 'bein' execution environment to run jobs in,</p>
  <p>* ``'job_or_dict'``: a 'Frontend' 'job' object, or a dictionary with 
  keys 'groups',</p>
  <p>* ``'fastq_root'``: path to raw fastq files,</p>
  <p>* ``'assembly_or_dict'``: an 'Assembly' object, or a dictionary of 
  'chromosomes' and 'index_path'.</p>
  <p>* ``'map_args'``: a dictionary of arguments passed to map_reads.</p>
  <p>Returns a dictionary with keys *group_id* from the job object and 
  values dictionaries mapping *run_id* to the corresponding return value of
  the 'map_reads' function.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="map_reads"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">map_reads</span>(<span class="sig-arg">ex</span>,
        <span class="sig-arg">fastq_file</span>,
        <span class="sig-arg">chromosomes</span>,
        <span class="sig-arg">bowtie_index</span>,
        <span class="sig-arg">maxhits</span>=<span class="sig-default">5</span>,
        <span class="sig-arg">antibody_enrichment</span>=<span class="sig-default">50</span>,
        <span class="sig-arg">name</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string"></code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">remove_pcr_duplicates</span>=<span class="sig-default">True</span>,
        <span class="sig-arg">bwt_args</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">via</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">lsf</code><code class="variable-quote">'</code></span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Runs ``bowtie`` in parallel over lsf for the `fastq_file` input. 
  Returns the full bamfile, its filtered version (see 
  'remove_duplicate_reads') and the mapping statistics dictionary (see 
  'bamstats').</p>
  <p>The input file will be split into subfiles if it contains more than 
  10M lines. The 'add_nh_flag' function will be called to add the number of
  hits per read in the bowtie output. If 'remove_pcr_duplicates' is *True*,
  the 'chromosomes' and 'maxhits' arguments are passed to the 
  'remove_duplicate_reads' function and the 'antibody_enrichment' will be 
  used as input to the 'poisson_threshold' function to compute its 
  'pilesize' argument.</p>
  <p>The mapping statistics dictionary is pickled and added to the 
  execution's repository, as well as both the full and filtered bam 
  files.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="parallel_density_sql"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">parallel_density_sql</span>(<span class="sig-arg">ex</span>,
        <span class="sig-arg">bamfile</span>,
        <span class="sig-arg">chromosomes</span>,
        <span class="sig-arg">nreads</span>=<span class="sig-default">1</span>,
        <span class="sig-arg">merge</span>=<span class="sig-default">-1</span>,
        <span class="sig-arg">read_extension</span>=<span class="sig-default">-1</span>,
        <span class="sig-arg">convert</span>=<span class="sig-default">True</span>,
        <span class="sig-arg">b2w_args</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">via</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">lsf</code><code class="variable-quote">'</code></span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Runs 'bam_to_density' for every chromosome in the 'chromosomes' 
  list.</p>
  <p>Generates 1 or 2 files depending if 'merge'&gt;=0 (shift and merge 
  strands into one track) or 'merge'&lt;0 (keep seperate tracks for each 
  strand) and returns their basename.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="parallel_density_wig"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">parallel_density_wig</span>(<span class="sig-arg">ex</span>,
        <span class="sig-arg">bamfile</span>,
        <span class="sig-arg">chromosomes</span>,
        <span class="sig-arg">nreads</span>=<span class="sig-default">1</span>,
        <span class="sig-arg">merge</span>=<span class="sig-default">-1</span>,
        <span class="sig-arg">read_extension</span>=<span class="sig-default">-1</span>,
        <span class="sig-arg">convert</span>=<span class="sig-default">True</span>,
        <span class="sig-arg">description</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string"></code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">alias</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">b2w_args</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">via</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">lsf</code><code class="variable-quote">'</code></span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Runs 'bam_to_density' in parallel for every chromosome in the 
  'chromosomes' list with 'sql' set to False. Returns a single text wig 
  file.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="poisson_threshold"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">poisson_threshold</span>(<span class="sig-arg">mu</span>,
        <span class="sig-arg">cutoff</span>=<span class="sig-default">0.95</span>,
        <span class="sig-arg">max_terms</span>=<span class="sig-default">100</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Calculate confidence threshold for Poisson distributions.</p>
  <p>Returns the largest integer *k* such that, for a Poisson distribution 
  random value X with mean 'mu', P(X &lt;= k) &lt;= 'cutoff'. It will 
  calculate no farther than k = 'max_terms'.  If it reaches that point, it 
  raises an exception.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="pprint_bamstats"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">pprint_bamstats</span>(<span class="sig-arg">sample_stats</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Pretty stdout-print for sample_stats.</p>
  <p>The input is the dictionary return by the ``bamstats`` call.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="remove_duplicate_reads"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">remove_duplicate_reads</span>(<span class="sig-arg">bamfile</span>,
        <span class="sig-arg">chromosomes</span>,
        <span class="sig-arg">maxhits</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">pilesize</span>=<span class="sig-default">1</span>,
        <span class="sig-arg">convert</span>=<span class="sig-default">False</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Filters a bam file for multi-hits above 'maxhits' and for duplicate 
  reads beyond 'pilesize'.</p>
  <p>Reads with NH tag &gt; maxhits are discarded, each genomic position 
  will have at most 'pilesize' reads per library and per strand. If the 
  'convert' flag is True, the reference sequence ids are replaced by their 
  names as provided in 'chromosomes'.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="bbcflib-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="http://bbcf.epfl.ch/bbcflib">bbcflib</a></th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Mon Aug 22 17:50:28 2011
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
