<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>bbcflib.genrep.GenRep</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="bbcflib-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="http://bbcf.epfl.ch/bbcflib">bbcflib</a></th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="bbcflib-module.html">Package&nbsp;bbcflib</a> ::
        <a href="bbcflib.genrep-module.html">Module&nbsp;genrep</a> ::
        Class&nbsp;GenRep
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="bbcflib.genrep.GenRep-class.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== CLASS DESCRIPTION ==================== -->
<h1 class="epydoc">Class GenRep</h1><p class="nomargin-top"></p>
<center>
<center>  <map id="uml_class_diagram_for_bbcflib__6" name="uml_class_diagram_for_bbcflib__6">
<area shape="rect" id="node5" href="bbcflib.genrep.GenRep-class.html#__init__" title="Create an object to query a GenRep repository." alt="" coords="17,39,697,57"/>
<area shape="rect" id="node5" href="bbcflib.genrep.GenRep-class.html#assemblies_available" title="Returns a list of assemblies available on genrep" alt="" coords="17,57,697,76"/>
<area shape="rect" id="node5" href="bbcflib.genrep.GenRep-class.html#assembly" title="Get an Assembly object corresponding to *assembly*." alt="" coords="17,76,697,95"/>
<area shape="rect" id="node5" href="bbcflib.genrep.GenRep-class.html#fasta_from_data" title="Get a fasta file with sequences corresponding to the features in the bed or sqlite file." alt="" coords="17,95,697,113"/>
<area shape="rect" id="node5" href="bbcflib.genrep.GenRep-class.html#fasta_path" title="Returns the path to the compressed fasta file, for the whole assembly or &#160;for a single chromosome." alt="" coords="17,113,697,132"/>
<area shape="rect" id="node5" href="bbcflib.genrep.GenRep-class.html#get_genrep_objects" title="Get a list of GenRep objets ..." alt="" coords="17,132,697,151"/>
<area shape="rect" id="node5" href="bbcflib.genrep.GenRep-class.html#get_sequence" title="Parses a slice request to the repository." alt="" coords="17,151,697,169"/>
<area shape="rect" id="node5" href="bbcflib.genrep.GenRep-class.html#get_sequence_straight" title="each element in input_list are lists containing: the chromosome name, the &#160;start and the end position." alt="" coords="17,169,697,188"/>
<area shape="rect" id="node5" href="bbcflib.genrep.GenRep-class.html#is_available" title="Returns a list of assemblies available on genrep" alt="" coords="17,188,697,207"/>
<area shape="rect" id="node5" href="bbcflib.genrep.GenRep-class.html#is_up" title="Check if genrep webservice is available" alt="" coords="17,207,697,225"/>
<area shape="rect" id="node5" href="bbcflib.genrep.GenRep-class.html#query_url" title="Assemble a URL to call *method* for *assembly* on the repository." alt="" coords="17,225,697,244"/>
<area shape="rect" id="node5" href="bbcflib.genrep.GenRep-class.html#statistics" title="Return (di&#45;)nucleotide counts or frequencies for an assembly, writes in file ``output`` if provided." alt="" coords="17,244,697,263"/>
<area shape="rect" id="node1" href="bbcflib.genrep.GenRep-class.html" title="bbcflib.genrep.GenRep" alt="" coords="5,6,709,269"/>
</map>
  <img src="uml_class_diagram_for_bbcflib__6.gif" alt='' usemap="#uml_class_diagram_for_bbcflib__6" ismap="ismap" class="graph-without-title" />
</center>
</center>
<hr />
<!-- ==================== INSTANCE METHODS ==================== -->
<a name="section-InstanceMethods"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Instance Methods</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-InstanceMethods"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.genrep.GenRep-class.html#__init__" class="summary-sig-name">__init__</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">url</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">http://bbcftools.vital-it.ch/genrep/</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">root</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">/db/genrep</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">intype</span>=<span class="summary-sig-default">0</span>,
        <span class="summary-sig-arg">config</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">section</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">genrep</code><code class="variable-quote">'</code></span>)</span><br />
      Create an object to query a GenRep repository.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="assemblies_available"></a><span class="summary-sig-name">assemblies_available</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Returns a list of assemblies available on genrep</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.genrep.GenRep-class.html#assembly" class="summary-sig-name">assembly</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">assembly</span>)</span><br />
      Get an Assembly object corresponding to *assembly*.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.genrep.GenRep-class.html#fasta_from_data" class="summary-sig-name">fasta_from_data</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">chromosomes</span>,
        <span class="summary-sig-arg">data_path</span>,
        <span class="summary-sig-arg">out</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">chunk</span>=<span class="summary-sig-default">50000</span>)</span><br />
      Get a fasta file with sequences corresponding to the features in the 
      bed or sqlite file.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="fasta_path"></a><span class="summary-sig-name">fasta_path</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">assembly</span>,
        <span class="summary-sig-arg">chromosome</span>=<span class="summary-sig-default">None</span>)</span><br />
      Returns the path to the compressed fasta file, for the whole assembly
      or for a single chromosome.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.genrep.GenRep-class.html#get_genrep_objects" class="summary-sig-name">get_genrep_objects</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">url_tag</span>,
        <span class="summary-sig-arg">info_tag</span>,
        <span class="summary-sig-arg">filters</span>=<span class="summary-sig-default">None</span>)</span><br />
      Get a list of GenRep objets ...</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_sequence"></a><span class="summary-sig-name">get_sequence</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">chr_id</span>,
        <span class="summary-sig-arg">coord_list</span>)</span><br />
      Parses a slice request to the repository.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_sequence_straight"></a><span class="summary-sig-name">get_sequence_straight</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">input_list</span>)</span><br />
      each element in input_list are lists containing: the chromosome name,
      the start and the end position.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="is_available"></a><span class="summary-sig-name">is_available</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">assembly</span>)</span><br />
      Returns a list of assemblies available on genrep</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="is_up"></a><span class="summary-sig-name">is_up</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Check if genrep webservice is available</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="query_url"></a><span class="summary-sig-name">query_url</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">method</span>,
        <span class="summary-sig-arg">assembly</span>)</span><br />
      Assemble a URL to call *method* for *assembly* on the repository.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.genrep.GenRep-class.html#statistics" class="summary-sig-name">statistics</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">assembly</span>,
        <span class="summary-sig-arg">output</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">frequency</span>=<span class="summary-sig-default">False</span>,
        <span class="summary-sig-arg">matrix_format</span>=<span class="summary-sig-default">False</span>)</span><br />
      Return (di-)nucleotide counts or frequencies for an assembly, writes in file ``output`` if provided.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
  <tr>
    <td colspan="2" class="summary">
    <p class="indent-wrapped-lines"><b>Inherited from <code>object</code></b>:
      <code>__delattr__</code>,
      <code>__format__</code>,
      <code>__getattribute__</code>,
      <code>__hash__</code>,
      <code>__new__</code>,
      <code>__reduce__</code>,
      <code>__reduce_ex__</code>,
      <code>__repr__</code>,
      <code>__setattr__</code>,
      <code>__sizeof__</code>,
      <code>__str__</code>,
      <code>__subclasshook__</code>
      </p>
    </td>
  </tr>
</table>
<!-- ==================== PROPERTIES ==================== -->
<a name="section-Properties"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Properties</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Properties"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
  <tr>
    <td colspan="2" class="summary">
    <p class="indent-wrapped-lines"><b>Inherited from <code>object</code></b>:
      <code>__class__</code>
      </p>
    </td>
  </tr>
</table>
<!-- ==================== METHOD DETAILS ==================== -->
<a name="section-MethodDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Method Details</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-MethodDetails"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
</table>
<a name="__init__"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">__init__</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">url</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">http://bbcftools.vital-it.ch/genrep/</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">root</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">/db/genrep</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">intype</span>=<span class="sig-default">0</span>,
        <span class="sig-arg">config</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">section</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">genrep</code><code class="variable-quote">'</code></span>)</span>
    <br /><em class="fname">(Constructor)</em>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Create an object to query a GenRep repository.</p>
  <p>GenRep is the in-house repository for sequence assemblies for the BBCF
  in Lausanne.  This is an object that wraps its use in Python in an 
  idiomatic way.</p>
  <p>Create a GenRep object with the base URL to the GenRep system, and the
  root path of GenRep's files.  For instance:</p>
<pre class="literalblock">
   g = GenRep('genrep.epfl.ch', '/path/to/genrep/indices')
</pre>
  <p>To get an assembly from the repository, call the assembly method with 
  either the integer assembly ID or the string assembly name.  This returns
  an Assembly object:</p>
<pre class="literalblock">
   a = g.assembly(3)
   b = g.assembly('mus')
</pre>
  <dl class="fields">
    <dt>Overrides:
        object.__init__
    </dt>
  </dl>
</td></tr></table>
</div>
<a name="assembly"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">assembly</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">assembly</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Get an Assembly object corresponding to *assembly*.</p>
  <p>*assembly* may be an integer giving the assembly ID, or a string 
  giving the assembly name.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="fasta_from_data"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">fasta_from_data</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">chromosomes</span>,
        <span class="sig-arg">data_path</span>,
        <span class="sig-arg">out</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">chunk</span>=<span class="sig-default">50000</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Get a fasta file with sequences corresponding to the features in the 
  bed or sqlite file.</p>
  <p>Returns the name of the file and the total sequence size.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_genrep_objects"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_genrep_objects</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">url_tag</span>,
        <span class="sig-arg">info_tag</span>,
        <span class="sig-arg">filters</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Get a list of GenRep objets ... attribute url_tag : the GenrepObject 
  type (plural) ... attribute info_tag : the GenrepObject type (singular) 
  Optionals attributes : ... attribute filters : a dict that is used to 
  filter the response from GenRep. Example : To get the genomes related to 
  'Mycobacterium leprae' species. First get the species with the right name
  : species = get_genrep_objects('organisms', 'organism', 
  {'species':'Mycobacterium leprae'})[0] genomes = 
  get_genrep_objects('genomes', 'genome', {'organism_id':species.id})</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="statistics"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">statistics</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">assembly</span>,
        <span class="sig-arg">output</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">frequency</span>=<span class="sig-default">False</span>,
        <span class="sig-arg">matrix_format</span>=<span class="sig-default">False</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <pre class="literalblock">

Return (di-)nucleotide counts or frequencies for an assembly, writes in file ``output`` if provided.
Example of result:
{
    &quot;TT&quot;: 13574667
    &quot;GG&quot;: 3344762
    &quot;CC&quot;: 3365555
    &quot;AA&quot;: 13571722
    &quot;A&quot;: 32370285
    &quot;TA&quot;: 6362526
    &quot;GT&quot;: 4841536
    &quot;AC&quot;: 4846697
    &quot;N&quot;: 0
    &quot;C&quot;: 17781115
    &quot;TC&quot;: 6228639
    &quot;GA&quot;: 6231575
    &quot;CG&quot;: 3131283
    &quot;GC: 3340219
    &quot;CT&quot;: 5079814
    &quot;AG&quot;: 5075950
    &quot;G&quot;: 17758095
    &quot;TG&quot;: 6206098
    &quot;CA&quot;: 6204462
    &quot;AT&quot;: 8875914
    &quot;T&quot;: 32371931
}
Total = A + T + G + C

if matrix_format is True output is for example:
 &gt;Assembly: sacCer2
1   0.309798640038793   0.308714120881750   0.190593944221299   0.190893294858157

</pre>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="bbcflib-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="http://bbcf.epfl.ch/bbcflib">bbcflib</a></th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Mon Aug 22 17:50:29 2011
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
