<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>bbcflib.rnaseq</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="bbcflib-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="http://bbcf.epfl.ch/bbcflib">bbcflib</a></th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="bbcflib-module.html">Package&nbsp;bbcflib</a> ::
        Module&nbsp;rnaseq
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="bbcflib.rnaseq-module.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== MODULE DESCRIPTION ==================== -->
<h1 class="epydoc">Module rnaseq</h1><p class="nomargin-top"></p>
<p>====================== Module: bbcflib.rnaseq 
  ======================</p>
  <p>No documentation</p>

<!-- ==================== CLASSES ==================== -->
<a name="section-Classes"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Classes</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Classes"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="bbcflib.rnaseq.AnnoteFinder-class.html" class="summary-name">AnnoteFinder</a><br />
      callback for matplotlib to display an annotation when points are 
        clicked on.
    </td>
  </tr>
</table>
<!-- ==================== FUNCTIONS ==================== -->
<a name="section-Functions"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Functions</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Functions"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.rnaseq-module.html#MAplot" class="summary-sig-name">MAplot</a>(<span class="summary-sig-arg">data</span>,
        <span class="summary-sig-arg">mode</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">interactive</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">deg</span>=<span class="summary-sig-default">4</span>,
        <span class="summary-sig-arg">bins</span>=<span class="summary-sig-default">30</span>,
        <span class="summary-sig-arg">alpha</span>=<span class="summary-sig-default">0.005</span>,
        <span class="summary-sig-arg">assembly_id</span>=<span class="summary-sig-default">None</span>)</span><br />
      Creates an &quot;MA-plot&quot; to compare transcription levels of a set of genes
in two different conditions.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="exons_labels"></a><span class="summary-sig-name">exons_labels</span>(<span class="summary-sig-arg">bamfile</span>)</span><br />
      List of the exons labels (SN: 'reference sequence name'; LN: 
      'sequence length') in *bamfile*.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.rnaseq-module.html#fetch_mappings" class="summary-sig-name">fetch_mappings</a>(<span class="summary-sig-arg">ex</span>,
        <span class="summary-sig-arg">path_or_assembly_id</span>)</span><br />
      Given an assembly ID, return a tuple (gene_ids, gene_names, 
      trans_to_gene_mapping, exons_to_trans_mapping).</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.rnaseq-module.html#inference" class="summary-sig-name">inference</a>(<span class="summary-sig-arg">cond1_label</span>,
        <span class="summary-sig-arg">cond1</span>,
        <span class="summary-sig-arg">cond2_label</span>,
        <span class="summary-sig-arg">cond2</span>,
        <span class="summary-sig-arg">method</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">normal</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">output</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">gene_names</span>=<span class="summary-sig-default">None</span>)</span><br />
      Runs DESeq comparing the counts in *cond1* and *cond2*.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="map_runs"></a><span class="summary-sig-name">map_runs</span>(<span class="summary-sig-arg">fun</span>,
        <span class="summary-sig-arg">runs</span>)</span><br />
      Parallelization of fun(run) executions</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.rnaseq-module.html#pairs_to_test" class="summary-sig-name">pairs_to_test</a>(<span class="summary-sig-arg">controls</span>)</span><br />
      *controls* is a dictionary of group_ids to True/False.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="pileup_file"></a><span class="summary-sig-name">pileup_file</span>(<span class="summary-sig-arg">bamfile</span>,
        <span class="summary-sig-arg">exons</span>)</span><br />
      Return a numpy array of the pileup of *bamfile* in order *exons*.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="bbcflib.rnaseq-module.html#rnaseq_workflow" class="summary-sig-name">rnaseq_workflow</a>(<span class="summary-sig-arg">ex</span>,
        <span class="summary-sig-arg">job</span>,
        <span class="summary-sig-arg">assembly</span>,
        <span class="summary-sig-arg">target</span>=<span class="summary-sig-default"><code class="variable-group">[</code><code class="variable-quote">'</code><code class="variable-string">genes</code><code class="variable-quote">'</code><code class="variable-group">]</code></span>,
        <span class="summary-sig-arg">via</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">lsf</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">output</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">maplot</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">normal</code><code class="variable-quote">'</code></span>)</span><br />
      Run RNASeq inference according to *job_info*.</td>
          <td align="right" valign="top">
            
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
</table>
<!-- ==================== VARIABLES ==================== -->
<a name="section-Variables"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Variables</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Variables"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="__package__"></a><span class="summary-name">__package__</span> = <code title="'bbcflib'"><code class="variable-quote">'</code><code class="variable-string">bbcflib</code><code class="variable-quote">'</code></code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="bam_to_sam"></a><span class="summary-name">bam_to_sam</span> = <code title="&lt;bein.program object at 0x10c04bc10&gt;">&lt;bein.program object at 0x10c04bc10&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="bowtie"></a><span class="summary-name">bowtie</span> = <code title="&lt;bein.program object at 0x10c04bad0&gt;">&lt;bein.program object at 0x10c04bad0&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="bowtie_build"></a><span class="summary-name">bowtie_build</span> = <code title="&lt;bein.program object at 0x10c04bb50&gt;">&lt;bein.program object at 0x10c04bb50&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="count_lines"></a><span class="summary-name">count_lines</span> = <code title="&lt;bein.program object at 0x10c04ba10&gt;">&lt;bein.program object at 0x10c04ba10&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="external_add_nh_flag"></a><span class="summary-name">external_add_nh_flag</span> = <code title="&lt;bein.program object at 0x10c04bb90&gt;">&lt;bein.program object at 0x10c04bb90&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="external_deseq"></a><span class="summary-name">external_deseq</span> = <code title="&lt;bein.program object at 0x1105798d0&gt;">&lt;bein.program object at 0x1105798d0&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="index_bam"></a><span class="summary-name">index_bam</span> = <code title="&lt;bein.program object at 0x10c04bd50&gt;">&lt;bein.program object at 0x10c04bd50&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="md5sum"></a><span class="summary-name">md5sum</span> = <code title="&lt;bein.program object at 0x10c04b990&gt;">&lt;bein.program object at 0x10c04b990&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="merge_bam"></a><span class="summary-name">merge_bam</span> = <code title="&lt;bein.program object at 0x10c04bd90&gt;">&lt;bein.program object at 0x10c04bd90&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="remove_lines_matching"></a><span class="summary-name">remove_lines_matching</span> = <code title="&lt;bein.program object at 0x10c04b950&gt;">&lt;bein.program object at 0x10c04b950&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="replace_bam_header"></a><span class="summary-name">replace_bam_header</span> = <code title="&lt;bein.program object at 0x10c04bc50&gt;">&lt;bein.program object at 0x10c04bc50&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="sam_to_bam"></a><span class="summary-name">sam_to_bam</span> = <code title="&lt;bein.program object at 0x10c04bbd0&gt;">&lt;bein.program object at 0x10c04bbd0&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="sleep"></a><span class="summary-name">sleep</span> = <code title="&lt;bein.program object at 0x10c04b9d0&gt;">&lt;bein.program object at 0x10c04b9d0&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="sort_bam"></a><span class="summary-name">sort_bam</span> = <code title="&lt;bein.program object at 0x10c04bc90&gt;">&lt;bein.program object at 0x10c04bc90&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="sort_bam_by_read"></a><span class="summary-name">sort_bam_by_read</span> = <code title="&lt;bein.program object at 0x10c04bcd0&gt;">&lt;bein.program object at 0x10c04bcd0&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="split_file"></a><span class="summary-name">split_file</span> = <code title="&lt;bein.program object at 0x10c04ba50&gt;">&lt;bein.program object at 0x10c04ba50&gt;</code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a name="touch"></a><span class="summary-name">touch</span> = <code title="&lt;bein.program object at 0x10c04b910&gt;">&lt;bein.program object at 0x10c04b910&gt;</code>
    </td>
  </tr>
</table>
<!-- ==================== FUNCTION DETAILS ==================== -->
<a name="section-FunctionDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Function Details</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-FunctionDetails"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
</table>
<a name="MAplot"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">MAplot</span>(<span class="sig-arg">data</span>,
        <span class="sig-arg">mode</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">interactive</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">deg</span>=<span class="sig-default">4</span>,
        <span class="sig-arg">bins</span>=<span class="sig-default">30</span>,
        <span class="sig-arg">alpha</span>=<span class="sig-default">0.005</span>,
        <span class="sig-arg">assembly_id</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <pre class="literalblock">

Creates an &quot;MA-plot&quot; to compare transcription levels of a set of genes
in two different conditions. It returns the name of the .png file produced,
and the name of a json containing enough information to reconstruct the plot using Javascript.

data:  rpy DataFrame object; two columns, each for a different condition.
mode:  display mode;
- if &quot;interactive&quot;, click on a point to display its name
- if &quot;normal&quot;, name of genes over 99%/under 1% quantile are displayed
alpha: a threshold for p-values: points beyond it are emphasized.
assembly_id  : if an assembly ID is given, the json output will provide links to information on genes.

</pre>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="fetch_mappings"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">fetch_mappings</span>(<span class="sig-arg">ex</span>,
        <span class="sig-arg">path_or_assembly_id</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Given an assembly ID, return a tuple (gene_ids, gene_names, 
  trans_to_gene_mapping, exons_to_trans_mapping). gene_ids is a list of 
  gene IDs; gene_names is a dictionary {gene ID: gene name}; 
  trans_to_gene_mapping is a dictionary {gene ID: [transcripts IDs]}; 
  exons_to_trans_mapping is a dictionary {transcript ID: [exons IDs]}.</p>
  <p>*path_or_assembly_id* can be a numeric or nominal ID for GenRep (e.g. 
  76 or 'hg19' for H.Sapiens), or a path to a file containing a JSON object
  which is read to get the mapping.</p>
  <p>Takes about 115 seconds to load.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="inference"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">inference</span>(<span class="sig-arg">cond1_label</span>,
        <span class="sig-arg">cond1</span>,
        <span class="sig-arg">cond2_label</span>,
        <span class="sig-arg">cond2</span>,
        <span class="sig-arg">method</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">normal</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">output</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">gene_names</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>Runs DESeq comparing the counts in *cond1* and *cond2*.</p>
  <p>Arguments: *cond1* and *cond2* are lists of numpy arrays. Each array 
  lists the number of reads mapping to a particular transcript. 
  *cond1_label* and *cond2_label* are string which will be used to identify
  the two conditions in R. *gene_names* is a dictionary {gene ID: gene 
  name}.</p>
  <p>``deseq_inference`` writes a tab delimited text file of the conditions
  to R, the first column being the transcript name, followed by one column 
  for each numpy array in *cond1*, then one column for each numpy array in 
  *cond2*.</p>
  <p>Then it calls DESeq in R, writes out the results in a new, randomly 
  named file, and returns that filename.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="pairs_to_test"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">pairs_to_test</span>(<span class="sig-arg">controls</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <p>*controls* is a dictionary of group_ids to True/False.</p>
  <p>If all the values are True or all the values are False, then it 
  returns all unique pairs of group IDs.  Otherwise it returns all 
  combinations of one group ID with value True and one with value 
  False.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="rnaseq_workflow"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">rnaseq_workflow</span>(<span class="sig-arg">ex</span>,
        <span class="sig-arg">job</span>,
        <span class="sig-arg">assembly</span>,
        <span class="sig-arg">target</span>=<span class="sig-default"><code class="variable-group">[</code><code class="variable-quote">'</code><code class="variable-string">genes</code><code class="variable-quote">'</code><code class="variable-group">]</code></span>,
        <span class="sig-arg">via</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">lsf</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">output</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">maplot</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">normal</code><code class="variable-quote">'</code></span>)</span>
  </h3>
  </td><td align="right" valign="top"
    >&nbsp;
    </td>
  </tr></table>
  
  <pre class="literalblock">
Run RNASeq inference according to *job_info*.

*output*: alternative name for output file. Otherwise it is random.
*maplot*: MA-plot of data.
- If 'interactive', one can click on a point (gene or exon) to display its name;
- if 'normal', name of genes over 99.9%/under 0.1% quantiles are displayed;
- if None, no figure is produced.
*target*: a string or array of strings indicating the targets to run inference (DESeq) on.
Targets can be 'genes', 'transcripts', 'exons'. E.g. ['genes','transcripts'], or 'exons'.
(This part of the workflow may fail if there are too few reads for DESeq to estimate variances.)
*job* is as Job object (or a dictionary of the same form) as returned from
HTSStation's frontend.

Whatever script calls this workflow needs to pass back the JSON string
it returns in some sensible way.  For the usual HTSStation
frontend, this just means printing it to stdout.

</pre>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="bbcflib-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="http://bbcf.epfl.ch/bbcflib">bbcflib</a></th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Mon Aug 22 17:50:28 2011
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
