<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>bbcflib: bbcflib::rnaseq Namespace Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javaScript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.7.1 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="namespaces.html"><span>Namespace&nbsp;List</span></a></li>
      <li><a href="namespacemembers.html"><span>Namespace&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="namespacebbcflib.html">bbcflib</a>      </li>
      <li><a class="el" href="namespacebbcflib_1_1rnaseq.html">rnaseq</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a> &#124;
<a href="#var-members">Variables</a>  </div>
  <div class="headertitle">
<h1>bbcflib::rnaseq Namespace Reference</h1>  </div>
</div>
<div class="contents">
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="func-members"></a>
Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#a975bfb6a1fe9beadc9f5eb3fdf2603ef">positive</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#ad7faf6b580da3cc1f71a5834aced2563">lsqnonneg</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#a6bd332f1d5322633cfb8248e9cab15b3">fetch_mappings</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#a65e85d3c9496c459318ff91b8e018522">fetch_labels</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#a521e2299d1fd927cddc3b46334ed563c">build_pileup</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#aa20ae70bccd2f689e892ac2caf9785f3">fusion</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#addb55a51916fc0585f2a3db2cb13fa36">save_results</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#a77cc2eafe75a6909cae99f4366739c37">genes_expression</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#a93cd4b073a999602a45d0170c8b761c4">transcripts_expression</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#a65206f5d600365077cb270ae797b50aa">estimate_size_factors</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a512ee8164003b5280c4cd01917746656"></a><!-- doxytag: member="bbcflib::rnaseq::to_rpkm" ref="a512ee8164003b5280c4cd01917746656" args="" -->
def&nbsp;</td><td class="memItemRight" valign="bottom"><b>to_rpkm</b></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#a28113694aeadb21bb19896b065c02695">rnaseq_workflow</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#a3d48ed91f262e4f47fcdd075193728c0">run_glm</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#ac41c6ef6fadb9739d1bb84c3b2a2d0b1">clean_before_deseq</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#aeb2683e6445cb01e3d1efab01554abf3">clean_deseq_output</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1rnaseq.html#add2f413a4b96bf081ee2fa3cb9401e1a">differential_analysis</a></td></tr>
<tr><td colspan="2"><h2><a name="var-members"></a>
Variables</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab01bb57e04e02640567036ff5ce64c64"></a><!-- doxytag: member="bbcflib::rnaseq::n" ref="ab01bb57e04e02640567036ff5ce64c64" args="" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><b>n</b></td></tr>
</table>
<hr/><a name="_details"></a><h2>Detailed Description</h2>
<div class="fragment"><pre class="fragment">
======================
Module: bbcflib.rnaseq
======================

Methods of the bbcflib's RNA-seq worflow. The main function is ``rnaseq_workflow()``,
and is usually called by bbcfutils' ``run_rnaseq.py``, e.g. command-line:

``python run_rnaseq.py -v local -c gapkowt.txt -d rnaseq -p transcripts,genes``

From a BAM file produced by an alignement on the **exonome**, gets counts of reads
on the exons, add them to get counts on genes, and uses least-squares to infer
counts on transcripts, avoiding to map on either genome or transcriptome.
Note that the resulting counts on transcripts are approximate.
The annotation of the bowtie index has to be consistent to that of the database (same assembly version).
</pre></div> <hr/><h2>Function Documentation</h2>
<a class="anchor" id="a521e2299d1fd927cddc3b46334ed563c"></a><!-- doxytag: member="bbcflib::rnaseq::build_pileup" ref="a521e2299d1fd927cddc3b46334ed563c" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::build_pileup </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>bamfile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>labels</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">From a BAM file, returns a dictionary of the form {feature ID: number of reads that mapped to it}.

:param bamfile: name of a BAM file.
:param labels: index references - as returned by **fetch_labels()**
:type bamfile: string
:type exons: list
</pre></div> 
</div>
</div>
<a class="anchor" id="ac41c6ef6fadb9739d1bb84c3b2a2d0b1"></a><!-- doxytag: member="bbcflib::rnaseq::clean_before_deseq" ref="ac41c6ef6fadb9739d1bb84c3b2a2d0b1" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::clean_before_deseq </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>filename</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Delete all lines of *filename* where counts are 0 in every run.</pre></div> 
</div>
</div>
<a class="anchor" id="aeb2683e6445cb01e3d1efab01554abf3"></a><!-- doxytag: member="bbcflib::rnaseq::clean_deseq_output" ref="aeb2683e6445cb01e3d1efab01554abf3" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::clean_deseq_output </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>filename</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Delete all lines of *filename* with NA's everywhere, add 0.5 to zero scores
before recalculating fold change, and remove rows' ids.</pre></div> 
</div>
</div>
<a class="anchor" id="add2f413a4b96bf081ee2fa3cb9401e1a"></a><!-- doxytag: member="bbcflib::rnaseq::differential_analysis" ref="add2f413a4b96bf081ee2fa3cb9401e1a" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::differential_analysis </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>ex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>result</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>rpath</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>design</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>contrast</em> = <code>None</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">For each file in *result*, launches an analysis of differential expression on the count
values, and saves the output in the MiniLIMS.

:param ex: the bein's execution.
:param result: dictionary {type:filename} as returned by rnaseq_workflow.
:param rpath: path to the R scripts (negbin.test.R).
:param design: name of the file containing the design matrix.
:param contrast: name of the file containing the contrasts matrix.
</pre></div> 
</div>
</div>
<a class="anchor" id="a65206f5d600365077cb270ae797b50aa"></a><!-- doxytag: member="bbcflib::rnaseq::estimate_size_factors" ref="a65206f5d600365077cb270ae797b50aa" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::estimate_size_factors </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>counts</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">
The total number of reads may be different between conditions or replicates.
This treatment makes different count sets being comparable. If rows are features
and columns are different conditions/replicates, each column is divided by the
geometric mean of the rows.
The median of these ratios is used as the size factor for this column. Size
factors may be used for further variance sabilization.

:param counts: an array of counts, each line representing a transcript, each
               column a different sample.
</pre></div> 
</div>
</div>
<a class="anchor" id="a65e85d3c9496c459318ff91b8e018522"></a><!-- doxytag: member="bbcflib::rnaseq::fetch_labels" ref="a65e85d3c9496c459318ff91b8e018522" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::fetch_labels </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>bamfile</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Returns a list of the exons/transcripts labels in the header of *bamfile*.</pre></div> 
</div>
</div>
<a class="anchor" id="a6bd332f1d5322633cfb8248e9cab15b3"></a><!-- doxytag: member="bbcflib::rnaseq::fetch_mappings" ref="a6bd332f1d5322633cfb8248e9cab15b3" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::fetch_mappings </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>assembly</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Given an assembly ID, returns a tuple
``(gene_mapping, transcript_mapping, exon_mapping, trans_in_gene, exons_in_trans)``

* [0] gene_mapping is a dict ``{gene ID: (gene name,start,end,length,chromosome)}``
* [1] transcript_mapping is a dictionary ``{transcript ID: (gene ID,start,end,length,chromosome)}``
* [2] exon_mapping is a dictionary ``{exon ID: ([transcript IDs],gene ID,start,end,chromosome)}``
* [3] trans_in_gene is a dict ``{gene ID: [IDs of the transcripts it contains]}``
* [4] exons_in_trans is a dict ``{transcript ID: [IDs of the exons it contains]}``

'Lenghts' are always the sum of the lengths of the exons in the gene/transcript.

:param path_or_assembly_id: can be a numeric or nominal ID for GenRep
(e.g. 11, 76 or 'hg19' for H.Sapiens), or a path to a file containing a
pickle object which is read to get the mapping.
</pre></div> 
</div>
</div>
<a class="anchor" id="aa20ae70bccd2f689e892ac2caf9785f3"></a><!-- doxytag: member="bbcflib::rnaseq::fusion" ref="aa20ae70bccd2f689e892ac2caf9785f3" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::fusion </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>X</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Takes a track-like generator with items of the form (chromosome,start,end,score)
and returns a merged track-like generator with summed scores.
This is to avoid having overlapping coordinates of features from both DNA strands,
which some genome browsers cannot handle for quantitative tracks.
</pre></div> 
</div>
</div>
<a class="anchor" id="a77cc2eafe75a6909cae99f4366739c37"></a><!-- doxytag: member="bbcflib::rnaseq::genes_expression" ref="a77cc2eafe75a6909cae99f4366739c37" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::genes_expression </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>exons_data</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>gene_mapping</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>exon_to_gene</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>ncond</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>nreads</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Get gene counts from exons counts.

Returns two dictionaries, one for counts and one for rpkm, of the form ``{gene ID: score}``.

:param exons_data: list of lists ``[exonsIDs,counts,rpkm,starts,ends,geneIDs,geneNames,strands,chrs]``.
:param exon_lengths: dictionary ``{exon ID: length}``
:param gene_mapping: dictionary ``{gene ID: (gene name,start,end,length,strand,chromosome)}``
:param exon_to_gene: dictionary ``{exon ID: gene ID}``.
:param ncond: number of samples.
</pre></div> 
</div>
</div>
<a class="anchor" id="ad7faf6b580da3cc1f71a5834aced2563"></a><!-- doxytag: member="bbcflib::rnaseq::lsqnonneg" ref="ad7faf6b580da3cc1f71a5834aced2563" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::lsqnonneg </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>C</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>x0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>tol</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>itmax_factor</em> = <code>3</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Linear least squares with nonnegativity constraints (NNLS), based on MATLAB's lsqnonneg function.

``(x,resnorm,res) = lsqnonneg(C,d)`` returns

* the vector *x* that minimizes norm(d-Cx) subject to x &gt;= 0
* the norm of residuals *resnorm* = norm(d-Cx)^2
* the residuals *res* = d-Cx

:param x0: Initial point for x.
:param tol: Tolerance to determine what is considered as close enough to zero.
:param itmax_factor: Maximum number of iterations.

:type C: nxm numpy array
:type d: nx1 numpy array
:type x0: mx1 numpy array
:type tol: float
:type itmax_factor: int
:rtype: *x*: numpy array, *resnorm*: float, *res*: numpy array

Reference: Lawson, C.L. and R.J. Hanson, Solving Least-Squares Problems, Prentice-Hall, Chapter 23, p. 161, 1974.
http://diffusion-mri.googlecode.com/svn/trunk/Python/lsqnonneg.py
</pre></div> 
</div>
</div>
<a class="anchor" id="a975bfb6a1fe9beadc9f5eb3fdf2603ef"></a><!-- doxytag: member="bbcflib::rnaseq::positive" ref="a975bfb6a1fe9beadc9f5eb3fdf2603ef" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::positive </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>x</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Set to zero all negative components of an array.</pre></div> 
</div>
</div>
<a class="anchor" id="a28113694aeadb21bb19896b065c02695"></a><!-- doxytag: member="bbcflib::rnaseq::rnaseq_workflow" ref="a28113694aeadb21bb19896b065c02695" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::rnaseq_workflow </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>ex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>job</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>bam_files</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pileup_level</em> = <code>[&quot;exons&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>genes</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>transcripts</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>via</em> = <code>&quot;lsf&quot;</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">
Main function of the workflow.

:rtype: None

:param ex: the bein's execution Id.
:param job: a Job object (or a dictionary of the same form) as returned from HTSStation's frontend.
:param bam_files: a complicated dictionary such as returned by mapseq.get_bam_wig_files.
:param pileup_level: a string or array of strings indicating the features you want to compare.
                     Targets can be 'genes', 'transcripts', or 'exons'.
:param via: 'local' or 'lsf'.
</pre></div> 
</div>
</div>
<a class="anchor" id="a3d48ed91f262e4f47fcdd075193728c0"></a><!-- doxytag: member="bbcflib::rnaseq::run_glm" ref="a3d48ed91f262e4f47fcdd075193728c0" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::run_glm </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>rpath</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>data_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>options</em> = <code>[]</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Run *rpath*/negbin.test.R on *data_file*.</pre></div> 
</div>
</div>
<a class="anchor" id="addb55a51916fc0585f2a3db2cb13fa36"></a><!-- doxytag: member="bbcflib::rnaseq::save_results" ref="addb55a51916fc0585f2a3db2cb13fa36" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::save_results </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>ex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>cols</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>conditions</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>group_ids</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>assembly</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>header</em> = <code>[]</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>feature_type</em> = <code>'features'</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Save results in a tab-delimited file, one line per feature, one column per run.

:param ex: bein's execution.
:param cols: list of iterables, each element being a column to write in the output.
:param conditions: list of strings corresponding to descriptions of the different samples.
:param group_ids: dictionary {group name: group ID}.
:param assembly: a GenRep Assembly object.
:param header: list of strings, the column headers of the output file.
:param feature_type: (str) the kind of feature of which you measure the expression.
</pre></div> 
</div>
</div>
<a class="anchor" id="a93cd4b073a999602a45d0170c8b761c4"></a><!-- doxytag: member="bbcflib::rnaseq::transcripts_expression" ref="a93cd4b073a999602a45d0170c8b761c4" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::rnaseq::transcripts_expression </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>exons_data</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>exon_lengths</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>transcript_mapping</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>trans_in_gene</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>exons_in_trans</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>ncond</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>nreads</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Get transcript rpkms from exon rpkms.

Returns two dictionaries, one for counts and one for rpkm, of the form ``{transcript ID: score}``.

:param exons_data: list of lists ``[exonsIDs,counts,rpkm,starts,ends,geneIDs,geneNames,strands,chrs]``
:param exon_lengths: dictionary ``{exon ID: length}``
:param transcript_mapping: dictionary ``{transcript ID: (gene ID,start,end,length,strand,chromosome)}``
:param trans_in_gene: dictionary ``{gene ID: [transcript IDs it contains]}``
:param exons_in_trans: dictionary ``{transcript ID: [exon IDs it contains]}``
:param ncond: number of samples
</pre></div> 
</div>
</div>
</div>
<!--- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&nbsp;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&nbsp;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&nbsp;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&nbsp;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr class="footer"/><address class="footer"><small>Generated on Fri Jul 13 2012 for bbcflib by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
