<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>bbcflib: bbcflib::mapseq Namespace Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javaScript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.7.1 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="namespaces.html"><span>Namespace&nbsp;List</span></a></li>
      <li><a href="namespacemembers.html"><span>Namespace&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="namespacebbcflib.html">bbcflib</a>      </li>
      <li><a class="el" href="namespacebbcflib_1_1mapseq.html">mapseq</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a> &#124;
<a href="#var-members">Variables</a>  </div>
  <div class="headertitle">
<h1>bbcflib::mapseq Namespace Reference</h1>  </div>
</div>
<div class="contents">
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="func-members"></a>
Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1mapseq.html#a8d7d8026e7942cfb95292e96d56c369c">fastq_dump</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1mapseq.html#a1cda9410e20c7718736ac6d332820a36">fastqc</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1mapseq.html#a5af1caf6e072306792c38f7d138daac7">run_fastqc</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebbcflib_1_1mapseq.html#aeed33d630a48857ee6181a74836697a2">get_fastq_files</a></td></tr>
<tr><td colspan="2"><h2><a name="var-members"></a>
Variables</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a767f5353b3dd8b6c4c11db3b88d5d0b3"></a><!-- doxytag: member="bbcflib::mapseq::demultiplex_path" ref="a767f5353b3dd8b6c4c11db3b88d5d0b3" args="" -->
string&nbsp;</td><td class="memItemRight" valign="bottom"><b>demultiplex_path</b> = &quot;/srv/demultiplexing/public/data/demultiplexing_minilims.files/&quot;</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a70569d5469bb3a5e643d4e10a706ae24"></a><!-- doxytag: member="bbcflib::mapseq::run_lib_name" ref="a70569d5469bb3a5e643d4e10a706ae24" args="" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><b>run_lib_name</b> = None</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9f0c05dd933005f49f1260c1595a7462"></a><!-- doxytag: member="bbcflib::mapseq::run" ref="a9f0c05dd933005f49f1260c1595a7462" args="" -->
tuple&nbsp;</td><td class="memItemRight" valign="bottom"><b>run</b> = str(run['url'])</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9e5b78df8a17d4ccf6346b0eb2dd04ca"></a><!-- doxytag: member="bbcflib::mapseq::dafl1" ref="a9e5b78df8a17d4ccf6346b0eb2dd04ca" args="" -->
list&nbsp;</td><td class="memItemRight" valign="bottom"><b>dafl1</b> = dafl[run['facility']]</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">tuple&nbsp;</td><td class="memItemRight" valign="bottom"><b>daf_data</b></td></tr>
</table>
<hr/><a name="_details"></a><h2>Detailed Description</h2>
<div class="fragment"><pre class="fragment">
======================
Module: bbcflib.mapseq
======================

This module provides functions useful to map raw reads using bowtie.
The most common workflow will first use ``map_reads`` which takes the following arguments:

  * ``'ex'``: an execution environment to run jobs in,

  * ``'fastq_file'``: the raw reads as a fastq file,

  * ``'chromosomes'``: a dictionary with keys 'chromosome_id' as used in the bowtie indexes and values a dictionary with usual chromosome names and their lengths,

  * ``'bowtie_index'``: the file prefix to the bowtie index,

  * ``'maxhits'``: the maximum number of hits per read to keep in the output (default *5*),

  * ``'antibody_enrichment'``: an approximate enrichement ratio of protein bound sequences over controls (default *50*),

  * ``'name'``: sample name to be used in descriptions and file names,

  * ``'remove_pcr_duplicates'``: whether to remove probable PCR amplification artifacts based on a Poisson confidence threshold (default *True*),

  * ``'bwt_args'``: a list of specific arguments for bowtie (in addition to ["-Sam",str(max(20,maxhits)),"--best","--strata"]).

The function ``map_groups`` will take a collection of sample as described in a *job* object from the ``frontend`` module and run fetch fastq files for each of them through using a ``daflims`` object, use an 'Assembly' object from ``genrep`` to get the bowtie indexes and run ``map_reads`` for each sample.

The second step in the workflow consists in generating genome-wide density maps with ``densities_groups`` which needs a 'bein' execution environment, the same *job* and *assembly* as above and a dictionary 'file_dict' as returned by ``map_groups``. The function then runs ``parallel_density_sql`` on each 'bam' file to obtain a normalized read density profile as a 'sqlite' file.

Below is the script used by the frontend::

from bbcflib import daflims, genrep, frontend, gdv, common
from bbcflib.mapseq import *
M = MiniLIMS( limspath )
working_dir = '/path/to/scratch/on/cluster'
hts_key = 'test_key'
gl = { 'hts_mapseq': {'url': 'http://htsstation.epfl.ch/mapseq/'},
       'genrep_url': 'http://bbcftools.vital-it.ch/genrep/',
       'bwt_root': '/db/genrep/',
       'script_path': '/srv/chipseq/lib',
       'lims': {'user': 'alice',
                 'passwd': {'lgtf': 'bob123',
                            'gva': 'joe456'}},
        'gdv': {'url': 'http://svitsrv25.epfl.ch/gdv',
                'email': 'alice.ecila@somewhere.edu',
                'key': 'xxxxxxxxxxxxxxxxxxxxxxxx'} }
assembly_id = 'mm9'
htss = frontend.Frontend( url=gl['hts_mapseq']['url'] )
job = htss.job( hts_key )
g_rep = genrep.GenRep( gl['genrep_url'], gl['bwt_root'] )
assembly = genrep.Assembly( assembly=assembly_id, genrep=g_rep )
daflims1 = dict((loc,daflims.DAFLIMS( username=gl['lims']['user'],
                                      password=gl['lims']['passwd'][loc] ))
                for loc in gl['lims']['passwd'].keys())
with execution( M, description=hts_key, remote_working_directory=working_dir ) as ex:
    job = get_fastq_files( ex, job, daflims1 )
    run_fastqc( ex, job, via=via )
    mapped_files = map_groups( ex, job, assembly )
    pdf = add_pdf_stats( ex, mapped_files,
                         dict((k,v['name']) for k,v in job.groups.iteritems()),
                         gl['script_path'] )
    density_files = densities_groups( ex, job, mapped_files, assembly.chromosomes )
    gdv_project = gdv.new_project( gl['gdv']['email'], gl['gdv']['key'],
                                   job.description, assembly.id, gl['gdv']['url'] )
    add_pickle( ex, gdv_project, description='py:gdv_json' )
print ex.id
allfiles = get_files( ex.id, M )
print allfiles
</pre></div> <hr/><h2>Function Documentation</h2>
<a class="anchor" id="a8d7d8026e7942cfb95292e96d56c369c"></a><!-- doxytag: member="bbcflib::mapseq::fastq_dump" ref="a8d7d8026e7942cfb95292e96d56c369c" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::mapseq::fastq_dump </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>options</em> = <code>None</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Binds ``fastq-dump`` to convert *sra* (short reads archive) to *fastq* format.
</pre></div> 
</div>
</div>
<a class="anchor" id="a1cda9410e20c7718736ac6d332820a36"></a><!-- doxytag: member="bbcflib::mapseq::fastqc" ref="a1cda9410e20c7718736ac6d332820a36" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::mapseq::fastqc </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>fastqfile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>outdir</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>options</em> = <code>None</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Binds ``fastqc`` (http://www.bioinformatics.bbsrc.ac.uk/) which generates a QC report of short reads present into the fastq file.
</pre></div> 
</div>
</div>
<a class="anchor" id="aeed33d630a48857ee6181a74836697a2"></a><!-- doxytag: member="bbcflib::mapseq::get_fastq_files" ref="aeed33d630a48857ee6181a74836697a2" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::mapseq::get_fastq_files </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>ex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>job</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>dafl</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>set_seed_length</em> = <code>True</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">
Will replace file references by actual file paths in the 'job' object.
These references are either 'dafl' run descriptions or urls.
Argument 'dafl' is a dictionary of 'Daflims' objects (keys are the facility names).
If 'set_seed_length' is true, a dictionary job.groups[gid]['seed_lengths']
of seed lengths is constructed with values corresponding to 70% of the
read length.
</pre></div> 
</div>
</div>
<a class="anchor" id="a5af1caf6e072306792c38f7d138daac7"></a><!-- doxytag: member="bbcflib::mapseq::run_fastqc" ref="a5af1caf6e072306792c38f7d138daac7" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def bbcflib::mapseq::run_fastqc </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>ex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>job</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>via</em> = <code>'lsf'</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">
Returns the name of the report file.
</pre></div> 
</div>
</div>
<hr/><h2>Variable Documentation</h2>
<a class="anchor" id="a7150e83ce31309b95932f47bcb9c1228"></a><!-- doxytag: member="bbcflib::mapseq::daf_data" ref="a7150e83ce31309b95932f47bcb9c1228" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">tuple bbcflib::mapseq::daf_data</td>
        </tr>
      </table>
</div>
<div class="memdoc">
<b>Initial value:</b><div class="fragment"><pre class="fragment"><a name="l00001"></a><a class="code" href="namespacebbcflib_1_1mapseq.html">00001</a> dafl1.fetch_fastq( str(run[<span class="stringliteral">&#39;facility&#39;</span>]), str(run[<span class="stringliteral">&#39;machine&#39;</span>]), run[<span class="stringliteral">&#39;run&#39;</span>], run[<span class="stringliteral">&#39;lane&#39;</span>],
<a name="l00002"></a>00002                                               libname=run.get(<span class="stringliteral">&#39;sequencing_library&#39;</span>) )
</pre></div>
</div>
</div>
</div>
<!--- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&nbsp;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&nbsp;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&nbsp;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&nbsp;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr class="footer"/><address class="footer"><small>Generated on Fri Jul 13 2012 for bbcflib by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
